aten/src/ATen/native/cuda/Shape.cu

```diff
@@ -0,0 +1,399 @@
+{
+  TORCH_CHECK((dim >= 0) && (dim < self.dim()),
+      "dimension ", dim, " out of range");
+  return self.stride(dim);
+}
+
+void check_shape_except_dim(const Tensor &first, const Tensor &second,
+                            int dimension)
+{
+  int first_dims = first.dim();
+  int second_dims = second.dim();
+  TORCH_CHECK(first_dims == second_dims,
+      "Tensors must have same number of dimensions: got ", first_dims,
+      " and ", second_dims);
+  for (int dim = 0; dim < first_dims; dim++) {
+    if (dim == dimension) {
+      continue;
+    }
+    int64_t first_dim_size = at::native::size(first, dim);
+    int64_t second_dim_size = at::native::size(second, dim);
+    TORCH_CHECK(first_dim_size == second_dim_size,
+        "Sizes of tensors must match except in dimension ", dim, ". Got ",
+        static_cast<long long>(first_dim_size), " and ",
+        static_cast<long long>(second_dim_size));
+  }
+}
+
+template <typename scalar_t>
+void parallel_cat(Tensor &out, THCState *state, const TensorList &inputs,
+    int64_t dimension, int nDims) {
+  // First, let's set up our kernel parameters. We start with a raw pointer to
+  // the storage for the output Tensor.
+  scalar_t *data = out.data_ptr<scalar_t>();
+
+  // Kernel Parameter
+  size_t tensorMetadataSize =
+    sizeof(CatArrInputTensor<scalar_t, unsigned int>) * CAT_ARRAY_BATCH_SIZE;
+  // TODO: replacement for THCudaMalloc()
+  auto d_inputs = static_cast<CatArrInputTensor<scalar_t, unsigned int> *>(
+      THCudaMalloc(state, tensorMetadataSize));
+
+  OutputTensorSizeStride<unsigned int, CAT_ARRAY_MAX_INPUT_DIMS> param;
+
+  // Next, let's initialize the size, stride arrays for the output Tensor.
+  for (int i = 0; i < nDims; ++i) {
+    param.outputSize[i] = dim_size(out, i);
+    param.outputStride[i] = dim_stride(out, i);
+  }
+
+  at::cuda::CUDAStream stream = at::cuda::getCurrentCUDAStream();
+
+  // Now we loop
+  int batchCounter = 0;
+  int64_t offset = 0;
+  for (int i = 0; i < inputs.size() ; i += CAT_ARRAY_BATCH_SIZE) {
+    // Re-allocate stackInputs every iteration to avoid read-after-write hazard
+    {
+      auto stackInputs_owner = THCudaHostAlloc(state, tensorMetadataSize);
+      CatArrInputTensor<scalar_t, unsigned int>* stackInputs =
+        static_cast<CatArrInputTensor<scalar_t, unsigned int>*>(
+          stackInputs_owner.get());
+      for (batchCounter = 0;
+           batchCounter < CAT_ARRAY_BATCH_SIZE &&
+             (i+batchCounter) < inputs.size();
+           ++batchCounter) {
+        int64_t dimSize = dim_size(inputs[i+batchCounter], dimension);
+
+        stackInputs[batchCounter].input =
+          inputs[i+batchCounter].data_ptr<scalar_t>();
+        stackInputs[batchCounter].offset = offset;
+        stackInputs[batchCounter].dimSize = dimSize;
+        stackInputs[batchCounter].nElements = inputs[i+batchCounter].numel();
+
+        // update offset
+        offset += dimSize;
+      }
+      THCudaCheck(cudaMemcpyAsync(
```

Comment:
since d_inputs and stackInputs will be just tensors in this case, you can use `copy_` with `non_blocking=true`